File: Android/src/jp/co/riso/android/text/InvalidCharacterFilter.java
/*
 * Copyright (c) 2014 RISO, Inc. All rights reserved.
 *
 * InvalidCharacterFilter.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */

package jp.co.riso.android.text;

import java.util.HashSet;
import java.util.Set;

import android.text.InputFilter;
import android.text.Spanned;

/**
 * @class InvalidCharacterFilter
 * 
 * @brief EditText filter for Invalid characters
 */
public class InvalidCharacterFilter implements InputFilter {
    private Set<Character> mCharSet;
    
    /**
     * @brief Constructor for invalid characters input filter
     * 
     * @param invalidChars Invalid characters in string format
     */
    public InvalidCharacterFilter(String invalidChars) {
        mCharSet = new HashSet<Character>();
        if (invalidChars == null) {
            return;
        }
        for (int i = 0; i < invalidChars.length(); i++) {
            mCharSet.add(invalidChars.charAt(i));
        }
    }

    @Override
    public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
        for (int i = start; i < end; i++) { 
            if (mCharSet.contains(source.charAt(i))) {
                return ""; 
            } 
        }
        return null; 
    }
}

File: Android/app/src/main/kotlin/jp/co/riso/android/text/InvalidCharacterFilter.kt
/*
 * Copyright (c) 2022 RISO, Inc. All rights reserved.
 *
 * InvalidCharacterFilter.kt
 * SmartDeviceApp
 * Created by: a-LINK Group
 */
package jp.co.riso.android.text

import android.text.InputFilter
import android.text.Spanned
import java.util.HashSet

/**
 * @class InvalidCharacterFilter
 *
 * @brief EditText filter for Invalid characters
 */
class InvalidCharacterFilter(invalidChars: String?) : InputFilter {
    private val _charSet: MutableSet<Char>
    override fun filter(
        source: CharSequence,
        start: Int,
        end: Int,
        dest: Spanned,
        dstart: Int,
        dend: Int
    ): CharSequence {
        for (i in start until end) {
            if (_charSet.contains(source[i])) {
                return ""
            }
        }
        return ""
    }

    /**
     * @brief Constructor for invalid characters input filter
     */
    init {
        _charSet = HashSet()
        invalidChars?.let {
            for (element in invalidChars) {
                _charSet.add(element)
            }
        }
    }
}